cmd/serverless-init/initcontainer/initcontainer.go

```diff
@@ -0,0 +1,116 @@
+package initcontainer
+
+import (
+	"context"
+	"fmt"
+	"os"
+	"os/exec"
+	"os/signal"
+	"strings"
+	"sync"
+	"syscall"
+	"time"
+
+	"github.com/DataDog/datadog-agent/cmd/serverless-init/log"
+	"github.com/DataDog/datadog-agent/cmd/serverless-init/metric"
+	"github.com/DataDog/datadog-agent/cmd/serverless-init/tag"
+	"github.com/DataDog/datadog-agent/pkg/logs"
+	"github.com/DataDog/datadog-agent/pkg/serverless/metrics"
+	"github.com/DataDog/datadog-agent/pkg/serverless/trace"
+)
+
+func Run(logConfig *log.Config, metricAgent *metrics.ServerlessMetricAgent, traceAgent *trace.ServerlessTraceAgent, args []string) {
+	log.Write(logConfig, []byte(fmt.Sprintf("[datadog init process] running cmd = >%v<", args)))
+	err := execute(logConfig, metricAgent, traceAgent, args)
+	if err != nil {
+		log.Write(logConfig, []byte(fmt.Sprintf("[datadog init process] exiting with code = %s", err)))
+	} else {
+		log.Write(logConfig, []byte("[datadog init process] exiting successfully"))
+	}
+}
+
+func execute(config *log.Config, metricAgent *metrics.ServerlessMetricAgent, traceAgent *trace.ServerlessTraceAgent, args []string) error {
+	sigs := make(chan os.Signal, 1)
+	defer close(sigs)
+	signal.Notify(sigs)
+	defer signal.Reset()
+	commandName, commandArgs := buildCommandParam(args[0])
+	cmd := exec.Command(commandName, commandArgs...)
+	cmd.Stdout = &log.CustomWriter{
+		LogConfig: config,
+	}
+	cmd.Stderr = &log.CustomWriter{
+		LogConfig: config,
+	}
+	cmd.SysProcAttr = &syscall.SysProcAttr{Setpgid: true}
+
+	go func() {
+		for sig := range sigs {
+			if sig != syscall.SIGURG {
+				log.Write(config, []byte(fmt.Sprintf("[datadog init process] %s received", sig)))
+			}
+			if sig != syscall.SIGCHLD {
+				if cmd.Process != nil {
+					_ = syscall.Kill(-cmd.Process.Pid, sig.(syscall.Signal))
+				}
+			}
+			if sig == syscall.SIGTERM {
+				metric.Shutdown(tag.GetBaseTags(), time.Now(), metricAgent.Demux)
+			}
+		}
+	}()
+	err := cmd.Start()
+	if err != nil {
+		return err
+	}
+	err = cmd.Wait()
+	if err != nil {
```

Comment:
Just so I understand the flow here, we wait for the process to complete, then we do one final flush if we had some kind of failure? What if err == nil? Why wouldn't we do a final flush then as well?

Reply:
Also, might make sense to have flushing be it's own function

Reply:
flushing is now in a separate method + final flush is done no matter what